packages/devextreme/js/__internal/data_controller.ts

```diff
@@ -0,0 +1,277 @@
+import { Deferred } from '@js/core/utils/deferred';
+
+import { extend } from '../core/utils/extend';
+import { isDefined } from '../core/utils/type';
+import ArrayStore from '../data/array_store';
+import { DataSource } from '../data/data_source/data_source';
+import { normalizeDataSourceOptions } from '../data/data_source/utils';
+
+interface DataSourceType {
+  _userData: unknown;
+  _pageSize: number;
+  dispose: () => void;
+  load: (loadOptions?: unknown) => Promise<unknown>;
+  loadSingle: (propName: string, propValue: unknown) => Promise<unknown>;
+  loadOptions: () => unknown;
+  cancel: (operationId: number) => void;
+  cancelAll: () => void;
+  filter: (filter?: unknown) => unknown;
+  group: (group?: unknown) => unknown;
+  items: () => unknown[];
+  key: () => string;
+  keyOf: (item: unknown) => string;
+  paginate: (paginate?: unknown) => unknown;
+  pageIndex: <TIdx = number | undefined>(pageIndex: TIdx) => TIdx extends number
+    ? undefined
+    : number;
```

Comment:
minor (can be skipped, just stylistic change)

```suggestion
  pageIndex: {
    (): number;
    (pageIndex: number): void;
  };
```